src/config.rs

```diff
@@ -0,0 +1,175 @@
+//! Config trait for types that can be configured from sources
+
+use crate::{Error, Result, Source};
+use serde::de::DeserializeOwned;
+use std::str::FromStr;
+
+/// Trait for types that can be configured by visiting sources
+/// 
+/// The Config implementation visits sources to gather the values it needs
+/// and assembles them into the final configuration.
+pub trait Config: Sized {
+    /// Configure this type by visiting the provided sources
+    /// 
+    /// Sources are checked in order, with later sources overriding earlier ones.
+    /// The implementation should query only the fields it needs.
+    fn configure_from_sources(sources: &[Box<dyn Source>]) -> Result<Self>;
```

Comment:
Config should not bother itself with multiple sources.
Builder should call `Config` methods appropriately.